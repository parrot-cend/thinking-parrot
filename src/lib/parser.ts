import { ButtonSchema, FormItemSchema, OptionsSchema } from './types'
import { createElement, VNode } from './VNode'

function formItemParser(formItem: FormItemSchema): VNode {
  const prop = formItem.prop || {}
  return createElement(formItem.tag, { prop })
}
function buttonParser(button: ButtonSchema): VNode {
  const prop = {
    size: button.size,
    type: button.type
  }
  return createElement('el-button', { prop })
}

export default function parser(source: string): string {
  const root = createElement('template')
  const form = createElement('el-form')
  root.insertChild(form)
  let sourceObj: OptionsSchema
  try {
    sourceObj = JSON.parse(source)
  } catch (e) {
    throw new Error('invalid source json')
  }
  if (sourceObj) {
    Object.keys(sourceObj).map(key => {
      const formItemNodes = sourceObj[key].formItem.map(item => formItemParser(item))
      const buttonNodes = sourceObj[key].buttons.map(button => buttonParser(button))
      form.insertChild(formItemNodes).insertChild(buttonNodes)
    })
  }
  return root.toString()
}
