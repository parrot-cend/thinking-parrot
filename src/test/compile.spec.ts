import fs from 'fs'
import parser from '../lib/parser'

it('generate correct template string', () => {
  const input = {
    queryForm: {
      formItem: [
        {
          model: 'searchContent',
          label: '搜索内容',
          tag: 'el-input',
          prop: {
            type: 'text'
          }
        }
      ],
      buttons: [
        {
          type: 'primary',
          size: 'small',
          text: 'Click me'
        }
      ]
    }
  }
  const output = parser(JSON.stringify(input))
  fs.writeFileSync('output.vue', output)
  expect(output).toMatchSnapshot()
})
