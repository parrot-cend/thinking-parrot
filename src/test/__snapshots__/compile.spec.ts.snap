// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate correct template string 1`] = `
"<template>
  <el-form>
    <el-input type=\\"text\\"></el-input>
    <el-button size=\\"small\\" type=\\"primary\\"></el-button>
  </el-form>
</template>;
"
`;
